File: src/main/java/org/intense/Ast/SetNode.java
package org.intense.Ast;

import org.intense.Environment;

public class SetNode extends ASTNode{
    AtomNode variableName;
    ASTNode expr;
    public SetNode(AtomNode atomNode, ASTNode parse) {
        this.variableName = atomNode;
        this.expr  = parse;
    }

    @Override
    public Object eval(Environment env) {
        // (set! my-heart "like this marvelous ")
        env.define(variableName.value,expr);
        return "set!";
    }
}

File: src/main/java/org/intense/ast/SetNode.kt
package org.intense.ast;

import org.intense.Env
import org.intense.Types.StrVal
import org.intense.Types.Value

class SetNode(atom: AtomNode, dataListNode: DataListNode) : ASTNode() {
    private var body: DataListNode? = dataListNode
    private var name: AtomNode = atom

    override fun eval(env: Env): Value {
        env.define(name.eval(env).toString(), body?.eval(env))
        return StrVal("set!")
    }
}
